### PR TITLE
rust: port `liballoc` to release `1.55.0`

### PR DESCRIPTION
**Pull request by AltF02**
Signed-off-by: Matthew Bakhtiari <dev@mtbk.me>

### CODE REVIEW

File: rust/alloc/sync.rs [hunk 1]
```diff
@@ -2334,6 +2341,20 @@ impl<T: ?Sized + Hash> Hash for Arc<T> {
...
+        /// calling `Arc::new(t)`.
+        ///
+        /// # Example
+        /// ```rust
+        /// # use std::sync::Arc;
+        /// let x = 5;
+        /// let arc = Arc::new(5);
+        ///
+        /// assert_eq!(Arc::from(x), arc);
+        /// ```
```
Comment (bjorn3) [inline rust/alloc/sync.rs]: Are you sure upstream has this? Doc comments inside function bodies give a warning AFAIK.

Comment (AltF02) [reply rust/alloc/sync.rs]: Oops, good catch its supposed to be outside the function

**Comment by ojeda** [comment]
Hmm... I have done it in my side to verify, and found also some missing differences from e.g. `rust/alloc/borrow.rs`.

**Comment by AltF02** [comment]
I did it manually, so I could've missed something. Is there a better way to port? (Sorry newbie)

File: rust/alloc/sync.rs (hunk 1 above)
Comment (ojeda) [reply rust/alloc/sync.rs]: @AltF02 Was this done by hand? I recommend porting the changes into the commit that introduced the files and then rebasing them.

**Comment by AltF02** [comment]
I didn't port the change from `rust/alloc/borrow.rs` because it was a rustc diagnostic. Which I assumed wasn't needed

**Comment by ojeda** [comment]
> I did it manually, so I could've missed something.

That explains it ;P

> Is there a better way to port?

I would recommend porting the changes into the commit that introduced the files, so that you can simply overwrite the files without manually editing them, then using `add --patch` to double-check the changes one by one, and finally asking Git to `rebase` or `cherry-pick` the commit.

**Comment by ojeda** [comment]
> I didn't port the change from `rust/alloc/borrow.rs` because it was a rustc diagnostic. Which I assumed wasn't needed

Yeah, the idea is to minimize the diff w.r.t. upstream, so please port the comments and things like that too (e.g. `rust/alloc/collections/mod.rs`).

**Comment by AltF02** [comment]
Cheers, thank you for the explanation. Will make an commit that does it properly in a second :smile:

**Comment by ojeda** [comment]
No rush! :)

By the way, it would be nice if you could also briefly list in the commit messages any extra changes needed, e.g. the `#[cfg(not(no_global_oom_handling))]` for `Arc::try_pin`.

File: rust/alloc/rc.rs [hunk 2]
```diff
@@ -143,7 +143,7 @@
 //! ```
 //!
 //! If our requirements change, and we also need to be able to traverse from
-//! `Owner` to `Gadget`, we will run into problems. An [`Rc`] pointer from `Owner`
+//! `Owner` to `Gadget`, we will run into problems. An [`Rc`] pointer from `Owner`
```
Comment (AltF02) [inline rust/alloc/rc.rs]: I honestly don't know what changed here

**Comment by AltF02** [comment]
Looks like that modification was only needed because I missed something in the use statements

File: rust/alloc/rc.rs (hunk 2 above)
Comment (nbdd0121) [reply rust/alloc/rc.rs]: This is a nbsp (0xA0) to space (0x20) change.

File: rust/alloc/boxed.rs [hunk 3]
```diff
@@ -189,8 +189,6 @@ impl<T> Box<T> {
     /// ```
     #[cfg(not(no_global_oom_handling))]
     #[inline(always)]
-    #[doc(alias = "alloc")]
-    #[doc(alias = "malloc")]
```
Comment (nbdd0121) [inline rust/alloc/boxed.rs:193]: I think these aliases are deliberately added on ourside.

Comment (AltF02) [reply rust/alloc/boxed.rs:193]: I can unstage these changes, but it looked like they were removed between 1.54.0 and 1.55.0, presumably this commit https://github.com/rust-lang/rust/commit/618c805746dbbc874afa538f97fc5f643aa9f6dc

Comment (nbdd0121) [reply rust/alloc/boxed.rs:193]: Oh, my bad. We should remove them then.

**Comment by ojeda** [comment]
The `rust/alloc/collections/mod.rs` comment is still missing. Also, the whitespace change in `rc.rs` was marked as resolved, but it is in the changes (upstream has nbsp).

**Comment by ojeda** [comment]
Looks good now!

Thanks a lot! :)

**Comment by AltF02** [comment]
Cheers

**Comment by ojeda** [comment]
I merged a bit too soon -- `try_pin` is fallible, so we should allow it under `no_global_oom_handling` (so, instead, include `Pin`).

Do you want to send a quick PR?

**Comment by AltF02** [comment]
Will create PR in a short second, sorry was unavailable

**Comment by ojeda** [comment]
No need to apologize!

**Pull request merged**